src/NHibernate/Collection/Generic/SetHelpers/SetSnapShot.cs

```diff
@@ -1,109 +1,178 @@
 using System;
 using System.Collections;
 using System.Collections.Generic;
+#if NETCOREAPP2_0
+using System.Runtime.Serialization;
+using System.Threading;
+#endif
 
 namespace NHibernate.Collection.Generic.SetHelpers
 {
+#if NETFX || NETSTANDARD2_0
+	// TODO 6.0: Consider removing this class in case we upgrade to .NET 4.7.2 and NET Standard 2.1,
+	// which have HashSet<T>.TryGetValue
 	[Serializable]
-	internal class SetSnapShot<T> : ISetSnapshot<T>
+	internal class SetSnapShot<T> : ICollection<T>, IReadOnlyCollection<T>, ICollection
 	{
-		private readonly List<T> _elements;
-		public SetSnapShot()
-		{
-			_elements = new List<T>();
-		}
+		private readonly Dictionary<T, T> _values;
+		private bool _hasNull;
 
 		public SetSnapShot(int capacity)
 		{
-			_elements = new List<T>(capacity);
+			_values = new Dictionary<T, T>(capacity);
 		}
 
 		public SetSnapShot(IEnumerable<T> collection)
 		{
-			_elements = new List<T>(collection);
+			_values = new Dictionary<T, T>();
+			foreach (var item in collection)
+			{
+				if (item == null)
+				{
+					_hasNull = true;
+				}
+				else
+				{
+					_values.Add(item, item);
+				}
+			}
 		}
 
-		public IEnumerator<T> GetEnumerator()
+		public bool TryGetValue(T equalValue, out T actualValue)
 		{
-			return _elements.GetEnumerator();
+			if (equalValue != null)
+			{
+				return _values.TryGetValue(equalValue, out actualValue);
+			}
+
+			actualValue = default(T);
+			return _hasNull;
 		}
 
-		IEnumerator IEnumerable.GetEnumerator()
+		public IEnumerator<T> GetEnumerator()
 		{
-			return GetEnumerator();
+			if (_hasNull)
+			{
+				yield return default(T);
+			}
+
+			foreach (var item in _values.Keys)
+			{
+				yield return item;
+			}
 		}
 
 		public void Add(T item)
 		{
-			_elements.Add(item);
+			if (item == null)
+			{
+				_hasNull = true;
+				return;
+			}
+
+			_values.Add(item, item);
 		}
 
 		public void Clear()
 		{
-			throw new InvalidOperationException();
+			_values.Clear();
+			_hasNull = false;
 		}
 
 		public bool Contains(T item)
 		{
-			return _elements.Contains(item);
+			return item == null ? _hasNull : _values.ContainsKey(item);
 		}
 
 		public void CopyTo(T[] array, int arrayIndex)
 		{
-			_elements.CopyTo(array, arrayIndex);
+			_values.Keys.CopyTo(array, arrayIndex);
```

Comment:
We are lacking the `_hasNull` handling here, for correctness. For consistency with current `GetEnumerator()`, it should be considered as the first element to copy.
```suggestion
			if (_hasNull)
				array[arrayIndex] = default(T);
			_values.Keys.CopyTo(array, arrayIndex + (_hasNull ? 1 : 0));
```

Reply:
Do you know what null means here? It looks like some legacy code considering that `HashSet` which is used for storing elements can't have null values.

Reply:
I see nothing that prevents `HashSet<T>` to have a `null`. They accept `null`.

Reply:
Ok.. I might mistake it with dictionary key or something else. But our code explicitly ignores null values from adding:
https://github.com/nhibernate/nhibernate-core/blob/20a61d1ca05dd792cd72ad3a6cb76307c7cb5642/src/NHibernate/Collection/Generic/PersistentGenericSet.cs#L168

Reply:
This is a thing which was done in af769586 as an "unrelated" by the way change, which according to comment of that time was required due to a limitation of the Iesi `set` NHibernate was using. (The Iesi set was required by lack of set in the .Net framework prior to .Net 4.)

I think keeping this limitation is now undue, it prevents supporting a set of `int?` elements with a null inside it, by example, with no reason. But removing is out of the scope of this PR.

Reply:
> I think keeping this limitation is now undue, it prevents supporting a set of int? elements with a null inside it, by example, with no reason. But removing this limitation is out of the scope of this PR.

Null removing is consistent with Java version. And for one thing it would break many-to-many filtering (I removed null elements for list also not so long ago - #2181). So I don't think this behavior should change

Reply:
So to conclude: maybe this change can be removed but skipping nulls should stay for entity types. And now you suggested change seems won't be used and can't be tested..